VideoLocker/src/main/java/org/edx/mobile/http/OauthRefreshTokenAuthenticator.java

```diff
@@ -53,14 +56,30 @@ public OauthRefreshTokenAuthenticator(Context context) {
     public Request authenticate(Route route, final Response response) throws IOException {
         logger.warn(response.toString());
 
-        if (!isTokenExpired(response.peekBody(HttpStatus.OK).string())) {
-            return null;
-        }
-
         final AuthResponse currentAuth = loginPrefs.getCurrentAuth();
         if (null == currentAuth || null == currentAuth.refresh_token) {
             return null;
         }
+        String response_body = response.peekBody(200).string();
+
+        switch (getErrorCode(response_body)) {
+            case TOKEN_EXPIRED_ERROR_MESSAGE:
+                break;
+            case TOKEN_NONEXISTENT_ERROR_MESSAGE:
+            case TOKEN_INVALID_GRANT_ERROR_MESSAGE:
+                // Retry request with the current access_token if the original access_token used in
+                // request does not match the current access_token. This case can occur when
+                // asynchronous calls are made and are attempting to refresh the access_token where
+                // one call succeeds but the other fails. https://github.com/edx/edx-app-android/pull/834
+                if (!response.request().headers().get("Authorization").split(" ")[1].equals(currentAuth.access_token)) {
```

Comment:
I don't think you want the authenticator to be called again from the reauthentication call itself, do you? This would actually be an [issue][1] that would need to be addressed if we were using the same `OkHttpClient` instance. And since you're adding the "Authorization" header manually, you don't need the header injection interceptor either. But sure we can address this separately, as it doesn't seem to be directly related to this fix.

[1]: https://github.com/square/okhttp/issues/662